8304939: os::win32::exit_process_or_thread should be marked noreturn

TheShermanTanker opened the pull request:
On Windows we have a non-trivial function (exit_process_or_thread) that provides the implementation of various functions like os::die, os::abort, &etc. Those os functions are marked as noreturn, so this implementation helper should also be noreturn.

The current change does several things around exit_process_or_thread, which is moved out of the os::win32 class since it does not need access to anything in that class other than the Ept enum, which is also moved to os_windows.cpp as well. The signature is changed to void, since exit_process_or_thread's current return value is simply the exit code parameter that it is passed, and to qualify as noreturn it should not return its current value of int. The only usage of this return value is in thread_native_entry, and this usage can easily be replaced by returning the res local that exit_process_or_thread is passed. For this os::infinite_sleep takes the place of the return statement in exit_process_or_thread, to make sure it qualifies as noreturn.

Although not mentioned in the title, raise_fail_fast has also been fixed. RaiseFailFastException is not itself noreturn, so os::infinite_sleep has been added there as well, to make raise_fail_fast qualify as a noreturn method

All the changes mentioned above fixes any Undefined Behaviour on Windows with regards to noreturn, by making all code marked noreturn (os::die() os::exit() etc) never return. This is not a problem on other platforms since their implementations of os::abort or os::die etc already call noreturn methods.

This fix also fixes compilation on gcc windows, which warns about noreturn methods returning prior to this changeset

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8304939](https://bugs.openjdk.org/browse/JDK-8304939): os::win32::exit_process_or_thread should be marked noreturn (**Enhancement** - P4)


### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Kim Barrett](https://openjdk.org/census#kbarrett) (@kimbarrett - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16303/head:pull/16303` \
`$ git checkout pull/16303`

Update a local copy of the PR: \
`$ git checkout pull/16303` \
`$ git pull https://git.openjdk.org/jdk.git pull/16303/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16303`

View PR using the GUI difftool: \
`$ git pr show -t 16303`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16303.diff">https://git.openjdk.org/jdk/pull/16303.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16303#issuecomment-1774549611)

TheShermanTanker commented:
Bumping, anyone?

dholmes-ora commented:
On my to-do list ...

dholmes-ora commented:
> moved out of the os::win32

This doesn't seem to add any value. Seemed fine to me in os::win32 versus os class.

dholmes-ora commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -3920,9 +3930,9 @@ void os::naked_short_nanosleep(jlong ns) {
 void os::infinite_sleep() {
-  while (true) {    // sleep forever ...
+  sleep:            // sleep forever ...
     Sleep(100000);  // ... 100 seconds at a time
-  }
+  goto sleep;
```
Again there needs to be a very strong justification for doing it this way.

dholmes-ora commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -4105,7 +4115,7 @@ static BOOL CALLBACK init_crit_sect_call(PINIT_ONCE, PVOID pcrit_sect, PVOID*) {
 }
 
-int os::win32::exit_process_or_thread(Ept what, int exit_code) {
+static void exit_process_or_thread(Ept what, int code) {
```
exit_code was clearer and avoids changes below

TheShermanTanker replied:
Noted, will return the name to exit_code

dholmes-ora commented on src/hotspot/os/windows/vmError_windows.cpp:
```diff
@@ -68,7 +68,6 @@ void VMError::interrupt_reporting_thread() {}
-  RaiseFailFastException(static_cast<PEXCEPTION_RECORD>(exrecord),
-                         static_cast<PCONTEXT>(context),
-                         flags);
+  RaiseFailFastException(static_cast<PEXCEPTION_RECORD>(exrecord), static_cast<PCONTEXT>(context), flags);
```
Curious that MS do not declare this as no_return themselves.

dholmes-ora reviewed (changes requested):
Supporting no_return seems a reasonable thing to do.

General complaint is that there is too much incidental change in the PR that to me adds no value. 

Some specific objections below.

Thanks

TheShermanTanker commented on src/hotspot/os/posix/os_posix.cpp:
```diff
@@ -889,9 +889,9 @@ char* os::build_agent_function_name(const char *sym_name, const char *lib_name,
 void os::infinite_sleep() {
-  while (true) {    // sleep forever ...
+  sleep:            // sleep forever ...
     ::sleep(100);   // ... 100 seconds at a time
-  }
+  goto sleep;
```
ah, when I was searching for functions to implement the noreturn with and stumbled across this one, I thought it could do with a goto instead of a while true since the intent seemed to be clearer. I can revert this if need be

dholmes-ora replied:
Please revert - `while (true)` conveys the exact intent.

TheShermanTanker commented:
> > moved out of the os::win32
> 
> This doesn't seem to add any value. Seemed fine to me in os::win32 versus os class.

I noticed that the method wasn't used anywhere outside of os_windows.cpp, and since I was rewriting it to be noreturn, I figured that a little refactoring of scope (from os::win32 to os_windows.cpp file scope) could help here

dholmes-ora commented:
> I figured that a little refactoring of scope (from os::win32 to os_windows.cpp file scope) could help here

The very loose, not well followed, historical convention here is that the Windows specific os class contains the methods defined by os.hpp, while the implementation details go into the os::win32 class. In many cases the choice is somewhat arbitrary, but there should still be a good reason to move something around. Unnecessary refactoring just makes the PR harder to understand.

TheShermanTanker commented:
Addressed some of the review comments

Side note: Should the Style Guide only permit noreturn for void methods? It's Undefined Behaviour when applied to something that returns int for instance, such as exit_process_or_thread here (which I had to refactor to void)

dholmes-ora commented:
> Side note: Should the Style Guide only permit noreturn for void methods? It's Undefined Behaviour when applied to something that returns int for instance, such as exit_process_or_thread here (which I had to refactor to void)

I think it is implied that attributes should only be used in a way that is valid.

dholmes-ora commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -505,12 +505,21 @@ struct tm* os::gmtime_pd(const time_t* clock, struct tm* res) {
+enum Ept { EPT_THREAD, EPT_PROCESS, EPT_PROCESS_DIE };
+// Wrapper around _endthreadex(), exit() and _exit()
+[[noreturn]]
+static void exit_process_or_thread(Ept what, int code);
```
Nit: code -> exit_code

dholmes-ora reviewed (approved):
I think the refactoring could have been rolled back further, but okay.

Needs second review.

Thanks

TheShermanTanker commented:
Paging for @kimbarrett, since the original issue was created by him

dholmes-ora commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -558,7 +567,8 @@ unsigned __stdcall os::win32::thread_native_entry(void* t) {
   // let it proceed to exit normally
-  return (unsigned)os::win32::exit_process_or_thread(os::win32::EPT_THREAD, res);
+  exit_process_or_thread(EPT_THREAD, res);
+  return res;
```
It has to call `exit_process_or_thread` because of the exit bug. That means we will directly call `_endthreadex` instead of implicitly doing that by having the entry function return. The `return res` is redundant but keeps the compiler happy given the required declaration of the entry method.

TheShermanTanker replied:
I'm not too sure what to make of this, since I don't know what the exit bug is about (Also, the return res doesn't cause an issue on MSVC under any circumstance, and would only do so on gcc if thread_native_entry was marked noreturn, which it isn't). I simply changed the return value to keep the original semantics of the code unchanged, I guess I should take this to mean I should keep my current changes as is?

kimbarrett replied:
Yes, keep as is in this area.

kimbarrett reviewed (approved):
To speed along the process, I'm approving this even though I still want the pointless
forward declaration of `thread_native_entry` removed.  I'll trust you to do that without
needing a re-review by me.

TheShermanTanker commented:
> To speed along the process, I'm approving this even though I still want the pointless forward declaration of `thread_native_entry` removed. I'll trust you to do that without needing a re-review by me.

I'll do just that, but I have a couple of questions: What about the comment accompanying the declaration? And is os::infinite_sleep appropriate for this purpose? I couldn't seem to find anything else to make exit_process_or_thread nor raise_fail_fast noreturn qualified

TheShermanTanker commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -505,12 +505,21 @@ struct tm* os::gmtime_pd(const time_t* clock, struct tm* res) {
+
+// The handler passed to _beginthreadex().
+// Called with the associated Thread* as the argument.
+static unsigned __stdcall thread_native_entry(void*);
```
I understand, but what about the useful (at least to me) comment? Should I move it to the definition of the method?

kimbarrett replied:
Sure, move the comment.

TheShermanTanker commented:
> > I figured that a little refactoring of scope (from os::win32 to os_windows.cpp file scope) could help here
> 
> The very loose, not well followed, historical convention here is that the Windows specific os class contains the methods defined by os.hpp, while the implementation details go into the os::win32 class. In many cases the choice is somewhat arbitrary, but there should still be a good reason to move something around. Unnecessary refactoring just makes the PR harder to understand.

Just noticed something in this review comment - I didn't move the method into the os class, I moved it into os_windows.cpp file scope

TheShermanTanker commented:
@dholmes-ora @kimbarrett Please see if you are fine with the final set of changes

dholmes-ora reviewed (approved):
Still seems fine to me. Thanks.

kimbarrett reviewed (approved):
Looks good.

TheShermanTanker commented:
Thanks for the reviews!

/integrate
